doc/HDF-4.2-to-4.3-migration.md

```diff
@@ -0,0 +1,75 @@
+# HDF 4.2 to 4.3 Migration Guide
+
+For the past few decades, HDF4 has deployed almost all of the library's header
+files, both public and private. This deployment scheme makes even trivial
+refactoring and bug fixing difficult to accomplish without breaking binary
+compatibility when "private" headers change.
+
+HDF 4.3.0 will finally separate the public and private APIs in both the single-
+and multi-file libraries, allowing us to begin the process of cleaning up
+several decades of accumulated technical debt.
+
+## Changes in HDF 4.3.0
+
+* Private header files are no longer distributed. A complete list can be found below.
+* The error-handling macros have been removed from `herr.h`.
+* Most of `hbitio.h` has been moved to a private header file.
+* Most of `hfile.h` has been moved to a private header file.
+* `hdf.h` no longer pulls in the `hdfi.h` internal header file. See below for the implications of this.
+
+## Migration Steps
+
+Most users should not have to make any changes to their applications, as long
+as they only use the public API as documented in the HDF4 Reference Manual.
+The single- and multi-file API calls are all unchanged and we have preserved
+the legacy HDF 3 API (found in `df.h`).
+
+If you HAVE used any of the undocumented functionality packeded with the library
+(e.g., `linklist.h`'s linked lists), you will either have to pull those
+source files out of an older version of the library for incorporation into
+your own code, or stick with HDF 4.2.16. Note that we don't plan to have
+additional HDF 4.2.x maintenance releases.
```

Comment:
Maybe a 4.2.17 with only build system changes (GH binaries) for those who don't want to move to 4.3?

Reply:
I'm not going to commit to that in a migration guide